File: Server/src/logic/game/PlayerIterator.java
package logic.game;

import logic.player.Player;

import java.util.Iterator;
import java.util.List;

/**
 * Created by Aedan Smith.
 */

public class PlayerIterator implements Iterator<Player> {
    private List<Player> players;
    private int index;

    public PlayerIterator(List<Player> players){
        this.players = players;
    }

    @Override
    public boolean hasNext() {
        return !players.isEmpty();
    }

    @Override
    public Player next() {
        index++;
        if (index >= players.size())
            index = 0;
        return players.get(index);
    }

    public void remove(Player player){
        int playerIndex = players.indexOf(player);
        if (playerIndex <= index){
            index--;
        }
        players.remove(playerIndex);
    }
}

File: Server/src/logic/game/PlayerIterator.kt
package logic.game

import logic.player.Player

/**
 * Created by Aedan Smith.
 */

class PlayerIterator(private val players: MutableList<Player>) : Iterator<Player> {
    private var index: Int = 0

    override fun hasNext(): Boolean {
        return !players.isEmpty()
    }

    override fun next(): Player {
        index++
        if (index >= players.size)
            index = 0
        return players[index]
    }

    fun remove(player: Player) {
        val playerIndex = players.indexOf(player)
        if (playerIndex <= index) {
            index--
        }
        players.removeAt(playerIndex)
    }
}
